### PR TITLE
Added the Clearable abstraction and applied to Credentials and User Profile per #250

### PR DESCRIPTION
**Pull request by JacobASeverson**
A Clearable abstraction will be useful for a types such as Credentials and UserProfile that may be
storing sensitive information. This new abstraction will allow implementers to clear out any sensitive
or unneeded information once it is no longer needed after authentication.

* Added the Clearable type

* Added Clearable to Credentials and implemented/test it for all subtypes

* Added Clearable to UserProfile and implemented/tested for OAuth 2

### CODE REVIEW

File: pac4j-http/src/main/java/org/pac4j/http/credentials/TokenCredentials.java [hunk 1]
```diff
@@ -42,4 +42,11 @@ public String getToken() {
     public String toString() {
         return CommonHelper.toString(this.getClass(), "token", this.token, "clientName", getClientName());
     }
+
+    @Override
+    public void clear() {
+        this.token = null;
+        this.setClientName(null);
+        this.setUserProfile(null);
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/credentials/TokenCredentials.java]: If ever we think we need to clear the user profile, you should call the `clear` method of the user profile.

Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/credentials/TokenCredentials.java]: In fact, my suggestion might not be a good idea. The user profile saved into this credentials is passed this way between an `Authenticator` and a `ProfileCreator` to eventually return it.
So clearing the credentials would clear the user profile, which we may not want.

**Comment by leleuj** [comment]
It's an excellent work. Thanks again.

I'm questioning myself: what's the real idea behind clearing credentials / user profiles? Removing critical information from the memory to avoid these information from being stolen?

**Comment by leleuj** [comment]
Thinking a little more about this, I stay with the idea that the `clear` concept should target critical information enabling authentication replay from attackers: in that case, it targets username/password in credentials, but also access token or PGT in user profiles.

I also think that clearing credentials should be independent from clearing user profiles. So I would keep the pull request as is except that the https://github.com/pac4j/pac4j/blob/master/pac4j-cas/src/main/java/org/pac4j/cas/client/rest/HttpTGTProfile.java should also clear its `ticketGrantingTicketId` and the https://github.com/pac4j/pac4j/blob/master/pac4j-oidc/src/main/java/org/pac4j/oidc/profile/OidcProfile.java should also clear its `accessToken`.

Thanks

**Comment by JacobASeverson** [comment]
Apologies, I went on vacation for the weekend - I'll make the recommended changes either tonight or tomorrow. Thanks!

**Comment by leleuj** [comment]
No problem.

**Pull request merged**